### PR TITLE
[SPARK-28234][CORE][PYTHON]  Add python and JavaSparkContext support to get resources

### PR DESCRIPTION
**Pull request by tgravescs**
## What changes were proposed in this pull request?

Add python api support and JavaSparkContext support for resources().  I needed the JavaSparkContext support for it to properly translate into python with the py4j stuff.

## How was this patch tested?

Unit tests added and manually tested in local cluster mode and on yarn.

### CODE REVIEW

**Comment by tgravescs** [comment]
@mengxr @jiangxb1987 @Ngone51

File: python/pyspark/resourceinformation.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    The array of addresses are resource specific and its up to the user to interpret the address.
+
+    One example is GPUs, where the addresses would be the indices of the GPUs
+
+    @param name the name of the resource
+    @param addresses an array of strings describing the addresses of the resource
+    """
+
+    _name = None
+    _addresses = None
```
Comment (HyukjinKwon) [inline python/pyspark/resourceinformation.py]: Seems we don't need this..

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/context.py [hunk 2]
```diff
@@ -1105,6 +1106,18 @@ def getConf(self):
...
         return conf
 
+    def resources(self):
+        resources = {}
+        jresources = self._jsc.resources()
+        for x in jresources:
+            name = jresources[x].name()
+            jaddresses = jresources[x].addresses()
+            addrs = []
+            for addr in jaddresses:
```
Comment (BryanCutler) [inline python/pyspark/context.py]: a comprehension would be a little cleaner here, e.g. `addrs = [addr for addr in jaddresses]`

File: python/pyspark/resourceinformation.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    def __init__(self, name, addresses):
+        self._name = name
+        self._addresses = addresses
+
+    def name(self):
```
Comment (BryanCutler) [inline python/pyspark/resourceinformation.py]: maybe use the `@property` decorator for these?

**Review by BryanCutler — COMMENTED** [review]
Looks good, just had a couple style comments

File: python/pyspark/resourceinformation.py (hunk 3 above)
Comment (HyukjinKwon) [reply python/pyspark/resourceinformation.py]: +1. For addresses too to be consistent with Scala side.

**Comment by HyukjinKwon** [comment]
Yea, looks fine to me in general too.

File: python/pyspark/resourceinformation.py (hunk 1 above)
Comment (tgravescs) [reply python/pyspark/resourceinformation.py]: yep, I'll remove

File: python/pyspark/resourceinformation.py (hunk 3 above)
Comment (tgravescs) [reply python/pyspark/resourceinformation.py]: ah good idea, I'll update

File: python/pyspark/context.py (hunk 2 above)
Comment (tgravescs) [reply python/pyspark/context.py]: will update

**Comment by tgravescs** [comment]
thanks for the reviews and suggestions, I've updated.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**